compiler/rustc_mir/src/transform/instcombine.rs

```diff
@@ -87,6 +93,81 @@ impl OptimizationFinder<'b, 'tcx> {
+            let local_being_derefed = match op.place()?.as_ref() {
+                PlaceRef { local, projection: [ProjectionElem::Deref] } => Some(local),
+                _ => None,
+            }?;
+
+            let stmt_index = location.statement_index;
+            // look behind for statement that assigns the local from a address of operator
+            // 6 is chosen as a heuristic determined by seeing the number of times the optimization kicked in compiling rust std
+            let lower_index = stmt_index.saturating_sub(6);
+            let statements_to_look_in = self.body.basic_blocks()[location.block].statements
+                [lower_index..stmt_index]
+                .iter()
+                .rev();
+            for stmt in statements_to_look_in {
+                match &stmt.kind {
+                    // Exhaustive match on statements to detect conditions that warrant we bail out of the optimization.
+                    // by only looking applying the optimization for immutable references e.g _local_being_deref = &_1;
+                    // we only need to check that opaque statements
+                    // like inline asm are not in the statements which can still modify `local_being_derefed`
+                    // or the value of `*local_being_derefed`
+                    rustc_middle::mir::StatementKind::Assign(box (l, r)) => {
+                        if l.local == local_being_derefed {
+                            match r {
+                                // looking for immutable reference e.g _local_being_deref = &_1;
+                                Rvalue::Ref(
+                                    _,
+                                    // Only apply the optimization if it is an immutable borrow.
+                                    BorrowKind::Shared,
+                                    place_taken_address_of,
+                                ) => {
+                                    self.optimizations
+                                        .unneeded_deref
+                                        .insert(location, *place_taken_address_of);
+                                    return Some(());
+                                }
+
+                                // We found an assignment of `local_being_deref` that is not an immutable ref, e.g the following sequence
+                                // _2 = &_1;
+                                // _3 = &5
+                                // _2 = _3;  <-- this means it is no longer valid to replace the last statement with `_5 = _1;`
+                                // _5 = (*_2);
+                                _ => return None,
+                            }
+                        }
+                        // at this point we know that `local_being_derefed` is not being assigned to.
+                        // Since there is an immutable borrow on `local_being_derefed`, it is not possible to take a mutable reference to `local_being_derefed`
```

Comment:
```suggestion
                        // Since there is an immutable borrow on `local_being_derefed`,
                        // it is not possible to take a mutable reference to `local_being_derefed`
```